livekit-rtc/livekit/rtc/chat.py

```diff
@@ -0,0 +1,140 @@
+        """Send a chat message to the end user using LiveKit Chat Protocol.
+
+        Args:
+            message (str): the message to send
+
+        Returns:
+            ChatMessage: the message that was sent
+        """
+        msg = ChatMessage(
+            message=message,
+            is_local=True,
+            participant=self._lp,
+        )
+        await self._lp.publish_data(
+            payload=json.dumps(msg.asjsondict()),
+            kind=DataPacketKind.KIND_RELIABLE,
+            topic=_CHAT_TOPIC,
+        )
+        return msg
+
+    async def update_message(self, message: "ChatMessage"):
+        """Update a chat message that was previously sent.
+
+        If message.deleted is set to True, we'll signal to remote participants that the message
+        should be deleted.
+        """
+        await self._lp.publish_data(
+            payload=json.dumps(message.asjsondict()),
+            kind=DataPacketKind.KIND_RELIABLE,
+            topic=_CHAT_UPDATE_TOPIC,
+        )
+
+    def on_message(self, callback: Callable[["ChatMessage"], None]):
+        """Register a callback to be called when a chat message is received from the end user."""
+        self._callback = callback
+
+    def _on_data_received(self, dp: DataPacket):
+        # handle both new and updates the same way, as long as the ID is in there
+        # the user can decide how to replace the previous message
+        if dp.topic == _CHAT_TOPIC or dp.topic == _CHAT_UPDATE_TOPIC:
+            try:
+                parsed = json.loads(dp.data)
+                msg = ChatMessage.from_jsondict(parsed)
+                if dp.participant:
+                    msg.participant = dp.participant
+                if self._callback:
+                    self._callback(msg)
+            except Exception as e:
+                logging.warning("failed to parse chat message: %s", e, exc_info=e)
+
+
+@dataclass
+class ChatMessage:
+    message: str = None
+    id: str = field(default_factory=generate_random_base62)
+    timestamp: datetime = field(default_factory=datetime.now)
+    deleted: bool = field(default=False)
+
+    # These fields are not part of the wire protocol. They are here to provide
+    # context for the application.
+    participant: Optional[Participant] = None
+    is_local: bool = field(default=False)
+
+    @classmethod
+    def from_jsondict(cls, d: Dict[str, Any]) -> "ChatMessage":
+        # older version of the protocol didn't contain a message ID, so we'll create one
+        id = d.get("id") or generate_random_base62()
+        timestamp = datetime.now()
+        if d.get("timestamp"):
+            timestamp = datetime.fromtimestamp(d.get("timestamp") / 1000.0)
+        msg = cls(
+            id=id,
+            timestamp=timestamp,
+        )
+        msg.update_from_jsondict(d)
+        return msg
+
+    def update_from_jsondict(self, d: Dict[str, Any]) -> None:
+        self.message = d.get("message")
```

Comment:
if message is None, should we default to the current message?
d.get("message", self.message)

Reply:
no, the protocol does not support delta updates. it's valid for message to be updated to None.. in the case of deletion.